security-proxy/src/main/java/org/georchestra/security/LdapUserDetailsRequestHeaderProvider.java

```diff
@@ -39,163 +51,323 @@
+ * the datadirectory's {@code security-proxy/headers-mapping.properties} file.
+ * <p>
+ * The final set of request headers sent to each specific proxified application
+ * is the aggregation of default headers and service specific headers.
+ * <p>
+ * Header mappings that apply to a specific target service take precedence over
+ * default headers.
+ * <p>
+ * Service name matches the ones assigned in {@code targets-mapping.properties}.
+ * For example, for service {@code analytics},
+ * {@code targets-mappings.properties} contains
+ * {@code analytics=http://analytics:8080/analytics/}, and
+ * {@code headers-mappings.properties} may contain
+ * {@code analytics.sec-firstname=givenName}.
+ * 
+ * 
  * @author jeichar
+ * @see SecurityRequestHeaderProvider
  */
 public class LdapUserDetailsRequestHeaderProvider extends HeaderProvider {
 
+    private static final String CACHED_USERNAME_KEY = "security-proxy-cached-username";
+
+    private static final String CACHED_HEADERS_KEY = "security-proxy-cached-attrs";
+
     protected static final Log logger = LogFactory
             .getLog(LdapUserDetailsRequestHeaderProvider.class.getPackage().getName());
 
-    private LdapUserSearch _userSearch;
-    private Map<String, String> _headerMapping;
-    private Pattern pattern;
-    private String orgSearchBaseDN;
+    private final FilterBasedLdapUserSearch _userSearch;
+    private final Pattern orgSeachMemberOfPattern;
+    private final String orgSearchBaseDN;
+
+    /**
+     * Header mappings that apply to all services (i.e. have no service prefix in
+     * header-mappings.properties), for example: ({@code sec-email=mail}
+     */
+    Map<String, String> defaultMappinsg = ImmutableMap.of();
+    /**
+     * Header mappings that apply to a specific target service, by service name,
+     * where service name matches the ones assigned in
+     * {@code targets-mapping.properties}. For example, for service
+     * {@code analytics}, {@code targets-mappings.properties} contains
+     * {@code analytics=http://analytics:8080/analytics/}, and
+     * {@code headers-mappings.properties} may contain
+     * {@code analytics.sec-firstname=givenName}.
+     */
+    Map<String, Map<String, String>> perServiceMappings = ImmutableMap.of();
 
     @Autowired
-    private LdapTemplate ldapTemplate;
+    @VisibleForTesting
+    LdapTemplate ldapTemplate;
 
     @Autowired
     private GeorchestraConfiguration georchestraConfiguration;
 
-    public LdapUserDetailsRequestHeaderProvider(LdapUserSearch userSearch, String orgSearchBaseDN,
-            Map<String, String> headerMapping) {
+    public LdapUserDetailsRequestHeaderProvider(FilterBasedLdapUserSearch userSearch, String orgSearchBaseDN) {
         Assert.notNull(userSearch, "userSearch must not be null");
-        Assert.notNull(headerMapping, "headerMapping must not be null");
+        // Q: is it ok for orgSearchBaseDN to be null?
```

Comment:
if null, then the pattern below (orgSeachMemberOfPattern) will just be less accurate, right ?

Reply:
if `null` then the pattern will be `"([^=,]+)=([^=,]+),null.*"`, and won't match anything?

Reply:
I will just add a null check for the constructor to fail fast. a null orgSearchBaseDN will make an unusable pattern, and `buildStandardOrganizationHeaders` will build a dn like `cn=org,null`:
```
 String dn = "cn=" + orgCn + "," + this.orgSearchBaseDN;
```

Means null is never actually used or someone would have complained, besides being out of scope for this pr.

Reply:
Done: 920d8a1bbbf0400f03b0356b365b8b39e77042de